tools/testbench/topology_ipc4.c

```diff
@@ -0,0 +1,1304 @@
+			tp->fw[tp->output_file_index].pipeline_id = ctx->pipeline_id;
+			tp->output_file_index++;
+		}
+		tb_setup_widget_ipc_msg(comp_info);
+	} else {
+		file = (struct ipc4_file_module_cfg *)comp_info->ipc_payload;
+		file->config.mode = FILE_READ;
+		file->config.rate = tp->fs_in;
+		file->config.channels = tp->channels_in;
+		file->config.frame_fmt = tp->frame_fmt;
+		file->config.direction = dir;
+
+		comp_info->instance_id = tp->instance_ids[SND_SOC_TPLG_DAPM_DAI_IN]++;
+		comp_info->module_id = 0x9d;
+		if (tb_is_pipeline_enabled(tp, ctx->pipeline_id)) {
+			if (tp->input_file_index >= tp->input_file_num) {
+				fprintf(stderr, "error: not enough input files for dai\n");
+				return -EINVAL;
+			}
+			file->config.fn = tp->input_file[tp->input_file_index];
+			tp->fr[tp->input_file_index].id = comp_info->module_id;
+			tp->fr[tp->input_file_index].instance_id = comp_info->instance_id;
+			tp->fr[tp->input_file_index].pipeline_id = ctx->pipeline_id;
+			tp->input_file_index++;
+		}
+		tb_setup_widget_ipc_msg(comp_info);
+	}
+
+	return 0;
+}
+
+int tb_new_pga(struct testbench_prm *tp)
+{
+	struct tplg_context *ctx = &tp->tplg;
+	struct tplg_comp_info *comp_info = ctx->current_comp_info;
+	struct ipc4_peak_volume_config volume;
+	struct snd_soc_tplg_ctl_hdr *tplg_ctl;
+	int ret;
+
+	comp_info->ipc_size =
+		sizeof(struct ipc4_peak_volume_config) + sizeof(struct ipc4_base_module_cfg);
+	comp_info->ipc_payload = calloc(comp_info->ipc_size, 1);
+	if (!comp_info->ipc_payload)
+		return -ENOMEM;
+
+	/* FIXME: move this to when the widget is actually set up */
+	comp_info->instance_id = tp->instance_ids[SND_SOC_TPLG_DAPM_PGA]++;
+	comp_info->module_id = 0x6;
+
+	tplg_ctl = calloc(ctx->hdr->payload_size, 1);
+	if (!tplg_ctl) {
+		free(comp_info->ipc_payload);
+		return -ENOMEM;
+	}
+
+	ret = tplg_new_pga(ctx, &volume, sizeof(struct ipc4_peak_volume_config),
+			   tplg_ctl, ctx->hdr->payload_size);
+	if (ret < 0) {
+		fprintf(stderr, "error: failed to create PGA\n");
+		goto out;
+	}
+
+	/* copy volume data to ipc_payload */
+	memcpy(comp_info->ipc_payload + sizeof(struct ipc4_base_module_cfg),
+	       &volume, sizeof(struct ipc4_peak_volume_config));
+
+	/* skip kcontrols for now */
+	ret = tplg_create_controls(ctx, ctx->widget->num_kcontrols, tplg_ctl,
+				   ctx->hdr->payload_size, &volume);
+	if (ret < 0) {
+		fprintf(stderr, "error: loading controls\n");
+		goto out;
+	}
+
+	tb_setup_widget_ipc_msg(comp_info);
+	free(tplg_ctl);
+	return ret;
+
+out:
+	free(tplg_ctl);
+	free(comp_info->ipc_payload);
+	return ret;
+}
+
+int tb_new_process(struct testbench_prm *tp)
+{
+	struct tplg_context *ctx = &tp->tplg;
+	struct tplg_comp_info *comp_info = ctx->current_comp_info;
+	struct snd_soc_tplg_ctl_hdr *tplg_ctl;
+	int ret;
+
+	ret = tplg_parse_widget_audio_formats(ctx);
+	if (ret < 0)
+		return ret;
+
+	tplg_ctl = calloc(ctx->hdr->payload_size, 1);
+	if (!tplg_ctl) {
+		free(comp_info->ipc_payload);
```

Comment:
is this correct? You only allocate it at line 790 below

Reply:
Oops, code arrange gone bad ...

Some randomly failing alloc function, like 1/100 in sof would be useful for stress tests.